src/main/scala/org/spartanz/parserz/ParsersModule.scala

```diff
@@ -52,6 +75,8 @@ trait ParsersModule {
     private[parserz] case class Delay[SI, SO, E, A](delayed: () => Grammar[SI, SO, E, A]) extends Grammar[SI, SO, E, A]
     private[parserz] case class Tag[SI, SO, E, A](value: Grammar[SI, SO, E, A], tag: String) extends Grammar[SI, SO, E, A]
     private[parserz] case class Map[SI, SO, E, A, B](value: Grammar[SI, SO, E, A], to: A => E \/ B, from: B => E \/ A) extends Grammar[SI, SO, E, B]
+    private[parserz] case class MapS[SI, SO, E, A, B](value: Grammar[SI, SO, E, A], to: (SI, A) => (SO, E \/ B), from: (SI, B) => (SO, E \/ A)) extends Grammar[SI, SO, E, B]
```

Comment:
This one can be implemented in terms of `MapES`, right? Also, `Map` can be implemented in terms of `MapES`. 

Unless there is an advantage on the implementation side to separating these cases (and sometimes there is!), it may be best to support a relatively small set of operations, to make maintenance easier.

Reply:
@jdegoes 
Hi John! Thanks a lot for the review!

The reason I have several `Map`s is the way state type parameters `SI` and `SO` are defined.

I cannot implement `Map` in terms of `MapS` because `Map` captures functions like `A => B` or `A => Option[B]`, but `MapS` is about functions `(SI, A) => (SO, B)` etc. 
I cannot build such function from just `A => B` because I have no knowledge of how to go from `SI` to `SO` since those types are completely unrelated. 

Thus `Map` and `MapS` have to be handled at the implementation level, where `SI` and `SO` are restricted further.

`MapES` is there for convenient use case: when mapping functions are not dealing with state, but error manipulations require state changes. That is why `MapES` captures `SI => (SO, E)`. 
I faced the same problem of not knowing how to go from `SI` to `SO` when trying to express it in terms of `MapS` (unless asking for extra function from user, and that wasn't friendly), and eventually ended up with 3 different mapping constructs.

Reply:
Makes sense. 👍